lntest/itest/lnd_etcd_failover_test.go

```diff
@@ -0,0 +1,116 @@
+// +build kvdb_etcd
+
+package itest
+
+import (
+	"context"
+	"io/ioutil"
+	"time"
+
+	"github.com/btcsuite/btcutil"
+	"github.com/lightningnetwork/lnd/channeldb/kvdb"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/lnrpc/routerrpc"
+	"github.com/lightningnetwork/lnd/lntest"
+	"github.com/stretchr/testify/require"
+)
+
+func testEtcdFailover(net *lntest.NetworkHarness, ht *harnessTest) {
+	ctxb := context.Background()
+
+	tmpDir, err := ioutil.TempDir("", "etcd")
+	etcdCfg, cleanup, err := kvdb.StartEtcdTestBackend(
+		tmpDir, uint16(lntest.NextAvailablePort()),
+		uint16(lntest.NextAvailablePort()),
+	)
+
+	require.NoError(ht.t, err)
+	defer cleanup()
+
+	password := []byte("the quick brown fox jumps the lazy dog")
+	entropy := [16]byte{1, 2, 3}
+	stateless := false
+	cluster := true
+
+	carol1, _, _, err := net.NewNodeWithSeedEtcd(
+		"Carol-1", etcdCfg, password, entropy[:], stateless, cluster,
+	)
+	if err != nil {
+		ht.Fatalf("unable to start Carol-1: %v", err)
+	}
+
+	ctxt, _ := context.WithTimeout(ctxb, defaultTimeout)
+	info1, err := carol1.GetInfo(ctxt, &lnrpc.GetInfoRequest{})
+
+	ctxt, _ = context.WithTimeout(ctxb, defaultTimeout)
+	if err := net.ConnectNodes(ctxt, carol1, net.Alice); err != nil {
+		ht.Fatalf("unable to connect Carol to Alice: %v", err)
+	}
+
+	// Open a channel with 100k satoshis between Carol and Alice with Alice
+	// being the sole funder of the channel.
+	chanAmt := btcutil.Amount(100000)
+	ctxt, _ = context.WithTimeout(ctxb, channelOpenTimeout)
+	_ = openChannelAndAssert(
+		ctxt, ht, net, net.Alice, carol1,
+		lntest.OpenChannelParams{
+			Amt: chanAmt,
+		},
+	)
+
+	// At this point Carol-1 is the elected leader, while Carol-2 will wait
+	// to become the leader when Carol-1 stops.
+	carol2, err := net.NewNodeEtcd(
+		"Carol-2", etcdCfg, password, cluster, false,
+	)
+	if err != nil {
+		ht.Fatalf("unable to start Carol-2: %v", err)
+	}
+
+	amt := btcutil.Amount(1000)
+	payReqs, _, _, err := createPayReqs(carol1, amt, 2)
+	if err != nil {
+		ht.Fatalf("Carol-2 is unable to create payment requests: %v",
+			err)
+	}
+	sendAndAssertSuccess(ht, net.Alice, &routerrpc.SendPaymentRequest{
+		PaymentRequest: payReqs[0],
+		TimeoutSeconds: 60,
+		FeeLimitSat:    noFeeLimitMsat,
+	})
+
+	// Shut down Carol-1 and wait for Carol-2 to become the leader.
+	shutdownAndAssert(net, ht, carol1)
+	err = carol2.WaitUntilLeader(30 * time.Second)
+	if err != nil {
+		ht.Fatalf("Waiting for Carol-2 to become the leader failed: %v",
+			err)
+	}
+
```

Comment:
If we surface that extra leader information, we can also query it here to assert that Carol-2 views Carol-1 as leader, then that it reviews itself as leader after the initial fail over.

Reply:
From the node's point of view, it becomes the leader whenever the `Campaign` call returns without an error. With the newly exposed `Leader` function we can run third party observer elector that does not participate in the election. Changed the itests to include this and assert on the leader ID. PTAL

Reply:
This is now added using a third-party observer.